### ISSUE TITLE
How could I get the code for "SAGEAttn-vT" in the paper?

### ISSUE THREAD

**Issue by Ezio-csm** [thu-ml/SageAttention#87]
Notice that there is a int8 quant algorithm for all the QKV in the paper, but haven't seen it in the repo.

**Comment by jason-huang03** [comment]
You can use the version that uses int8 QK and fp8 PV. This offers better performance than int8 for QKV.

**Comment by Ezio-csm** [comment]
Sorry that tensor core don't support fp8 on A100. Would you please update the "SAGEAttn-vT" version soon?

**Comment by jt-zhang** [comment]
Hello. The accuracy performance of SAGEAttn-vT is terrible in some situations, so we do not recommend using that.

If you are still willing to try, you can refer to:

``` python
import os
import pytest, pickle
import torch
import triton
import triton.language as tl
import torch.nn.functional as F

@triton.jit
def _attn_fwd_inner(acc, l_i, m_i, q, q_scale, 
                    K_ptrs, K_scale_ptr, V_ptrs, v_scale, 
                    start_m, qk_scale,  
                    BLOCK_M: tl.constexpr, HEAD_DIM: tl.constexpr, BLOCK_N: tl.constexpr,  
                    STAGE: tl.constexpr, offs_m: tl.constexpr, offs_n: tl.constexpr,  
                    N_CTX: tl.constexpr, fp8_v: tl.constexpr):
    lo, hi = 0, N_CTX

    for start_n in range(lo, hi, BLOCK_N):
        start_n = tl.multiple_of(start_n, BLOCK_N)
        k = tl.load(K_ptrs, mask = offs_n[None, :] < (N_CTX - start_n))
        k_scale = tl.load(K_scale_ptr, mask = offs_n[None, :] < (N_CTX - start_n))
        qk = tl.dot(q, k).to(tl.float32) * q_scale * k_scale
        m_ij = tl.maximum(m_i, tl.max(qk, 1) * qk_scale)
        qk = qk * qk_scale - m_ij[:, None]
        p = tl.math.exp2(qk)
        l_ij = tl.sum(p, 1)
        alpha = tl.math.exp2(m_i - m_ij)
        l_i = l_i * alpha + l_ij
        acc = acc * alpha[:, None]
        v = tl.load(V_ptrs, mask = offs_n[:, None] < (N_CTX - start_n))
        p = (p*127).to(tl.int8)
        acc += tl.dot(p.to(tl.int8), v) * v_scale /127   
        m_i = m_ij
        K_ptrs += BLOCK_N * HEAD_DIM
        K_scale_ptr += BLOCK_N
        V_ptrs += BLOCK_N 
    return acc, l_i


@triton.jit
def _attn_fwd(Q, K, V, Q_scale, K_scale, V_scale, sm_scale, M, Out,  
              stride_qz, stride_qh, stride_qm, stride_qk,  
              stride_kz, stride_kh, stride_kn, stride_kk,  
              stride_vz, stride_vh, stride_vk, stride_vn,  
              stride_oz, stride_oh, stride_om, stride_on,  
              Z, H, N_CTX,  
              HEAD_DIM: tl.constexpr,  
              BLOCK_M: tl.constexpr,  
              BLOCK_N: tl.constexpr,  
              STAGE: tl.constexpr  
              ):
    start_m = tl.program_id(0)
    off_hz = tl.program_id(1)
    off_z = off_hz // H
    off_h = off_hz % H
    qvk_offset = off_z.to(tl.int64) * stride_qz + off_h.to(tl.int64) * stride_qh
    vk_offset = qvk_offset // stride_qm
    offs_m = start_m * BLOCK_M + tl.arange(0, BLOCK_M)
    offs_n = tl.arange(0, BLOCK_N)
    offs_k = tl.arange(0, HEAD_DIM)
    Q_ptrs = Q + qvk_offset + offs_m[:, None] * stride_qm + offs_k[None, :] * stride_qk
    Q_scale_ptr = Q_scale + vk_offset + offs_m[:, None]
    K_ptrs = K + qvk_offset + offs_k[:, None] + offs_n[None, :] * stride_kn
    K_scale_ptr = K_scale + vk_offset + offs_n[None, :]
    V_ptrs = V + qvk_offset + offs_n[:, None] + offs_k[None, :] * N_CTX
    O_block_ptr = Out + qvk_offset + offs_m[:, None] * stride_qm + offs_k[None, :] * stride_qk
    m_i = tl.zeros([BLOCK_M], dtype=tl.float32) - float("inf")
    l_i = tl.zeros([BLOCK_M], dtype=tl.float32) + 1.0
    acc = tl.zeros([BLOCK_M, HEAD_DIM], dtype=tl.float32)
    qk_scale = sm_scale
    qk_scale *= 1.44269504  
    q = tl.load(Q_ptrs, mask = offs_m[:, None] < N_CTX)
    q_scale = tl.load(Q_scale_ptr, mask = offs_m[:, None] < N_CTX) 
    V_scale_ptr = V_scale + off_hz * HEAD_DIM + offs_k[None, :] 
    v_scale = tl.load(V_scale_ptr)
    acc, l_i = _attn_fwd_inner(acc, l_i, m_i, q, q_scale, K_ptrs, K_scale_ptr, V_ptrs, v_scale, 
                                    start_m, qk_scale,  
                                    BLOCK_M, HEAD_DIM, BLOCK_N,  
                                    4 - STAGE, offs_m, offs_n, N_CTX, V.dtype.element_ty == tl.float8e4nv  
                                    )
    acc = acc / l_i[:, None]
    tl.store(O_block_ptr, acc.to(Out.type.element_ty), mask = (offs_m[:, None] < N_CTX))


class _attention(torch.autograd.Function):
    @staticmethod
    def forward(ctx, q, k, v, q_scale, k_scale, v_scale, causal, sm_scale):
        BLOCK_M = 128
        BLOCK_N = 64
        HEAD_DIM_Q, HEAD_DIM_K = q.shape[-1], k.shape[-1]
        HEAD_DIM_V = v.shape[-1]
        assert HEAD_DIM_Q == HEAD_DIM_K and HEAD_DIM_K == HEAD_DIM_V
        assert HEAD_DIM_K in {16, 32, 64, 128, 256}
        o = torch.empty_like(q, dtype=torch.float16)
        stage = 3 if causal else 1
        extra_kern_args = {}

    
        grid = (triton.cdiv(q.shape[2], BLOCK_M), q.shape[0] * q.shape[1], 1)
        M = torch.empty((q.shape[0], q.shape[1], q.shape[2]), device=q.device, dtype=torch.float32)
        _attn_fwd[grid](
            q, k, v, q_scale, k_scale, v_scale, sm_scale, M, o,  
            q.stride(0), q.stride(1), q.stride(2), q.stride(3),  
            k.stride(0), k.stride(1), k.stride(2), k.stride(3),  
            v.stride(0), v.stride(1), v.stride(2), v.stride(3),  
            o.stride(0), o.stride(1), o.stride(2), o.stride(3),  
            q.shape[0], q.shape[1],  
            N_CTX=q.shape[2],  
            BLOCK_M=BLOCK_M, BLOCK_N=BLOCK_N, HEAD_DIM=HEAD_DIM_K,  
            STAGE=stage,  
            num_warps=4,  
            num_stages=3,  
            **extra_kern_args)

        return o

attention = _attention.apply

import torch.nn.functional as F
def precision_metric(quant_o, fa2_o): 
        x, xx = quant_o.float(), fa2_o.float() 
        sim = F.cosine_similarity(x.reshape(1, -1), xx.reshape(1, -1)).item()
        l1 =   ( (x - xx).abs().sum() / xx.abs().sum() ).item()
        rmse = torch.sqrt(torch.mean((x -xx) ** 2)).item()
        print(f'Cossim: {sim:.6f}, L1: {l1:.6f}, RMSE:{rmse:.6f}\n')


def test_op():
    q = 0.25*(torch.randn((1, 4, 4096, 64), dtype=torch.float16, device="cuda")).to(torch.float16)
    k = 0.25*(torch.randn((1, 4, 4096, 64), dtype=torch.float16, device="cuda")).to(torch.float16)
    v = 0.25*(torch.randn((1, 4, 4096, 64), dtype=torch.float16, device="cuda")).to(torch.float16)

    torch_stand = torch.nn.functional.scaled_dot_product_attention(q, k, v)
    k -= k.mean(dim=-2, keepdim=True)

    q_scale = torch.max(q.to(torch.float32).abs(), dim=3, keepdim=True)[0] / 127 
    k_scale = torch.max(k.to(torch.float32).abs(), dim=3, keepdim=True)[0] / 127
    q = (q/q_scale).round().to(torch.int8)
    k = (k/k_scale).round().to(torch.int8)
    v_scale = torch.max(v.to(torch.float32).abs(), dim=2, keepdim=True)[0] / 127 
    v = v.permute(0, 1, 3, 2).contiguous()
    v = v.permute(0, 1, 3, 2)
    v = (v/v_scale).round().to(torch.int8)

    sm_scale = HEAD_DIM**-0.5
    tri_out = attention(q, k, v, q_scale, k_scale, v_scale, causal, sm_scale)
    precision_metric(tri_out, torch_stand)


HAS_FLASH = False
TORCH_HAS_FP8 = hasattr(torch, 'float8_e5m2')
BATCH, N_HEADS, HEAD_DIM = 4, 32, 64
configs = []
for mode in ["fwd"]:
    for causal in [False]:  
        if mode == "bwd" and not causal:
            continue
        configs.append(
            triton.testing.Benchmark(
                x_names=["N_CTX"],
                x_vals=[2**i for i in range(10, 16)],
                line_arg="provider",
                line_vals=(["triton-fp8"] if TORCH_HAS_FP8 else []) +
                (["flash"] if HAS_FLASH else []),
                line_names=(["Triton [FP8]"] if TORCH_HAS_FP8 else []) +
                (["Flash-2"] if HAS_FLASH else []),
                styles=[("red", "-"), ("blue", "-")],
                ylabel="ms",
                plot_name=f"fused-attention-batch{BATCH}-head{N_HEADS}-d{HEAD_DIM}-{mode}-causal={causal}",
                args={
                    "H": N_HEADS,
                    "BATCH": BATCH,
                    "HEAD_DIM": HEAD_DIM,
                    "mode": mode,
                    "causal": causal,
                },
            ))


@triton.testing.perf_report(configs)
def bench_flash_attention(BATCH, H, N_CTX, HEAD_DIM, causal, mode, provider, device="cuda"):
    assert mode in ["fwd", "bwd"]
    warmup = 25
    rep = 100
    dtype = torch.float16
    if "triton" in provider:
        q = torch.randn((BATCH, H, N_CTX, HEAD_DIM), dtype=dtype, device=device, requires_grad=True)
        k = torch.randn((BATCH, H, N_CTX, HEAD_DIM), dtype=dtype, device=device, requires_grad=True)
        v = torch.randn((BATCH, H, N_CTX, HEAD_DIM), dtype=dtype, device=device, requires_grad=True)

        q_scale = torch.max(q.to(torch.float32).abs(), dim=3, keepdim=True)[0] / 127 
        k_scale = torch.max(k.to(torch.float32).abs(), dim=3, keepdim=True)[0] / 127
        q = (q/q_scale).round().to(torch.int8)
        k = (k/k_scale).round().to(torch.int8)
        v_scale = torch.max(v.to(torch.float32).abs(), dim=2, keepdim=True)[0] / 127 
        v = v.permute(0, 1, 3, 2).contiguous()
        v = v.permute(0, 1, 3, 2)
        v = (v/v_scale).to(torch.int8)
        sm_scale = 1.3
        fn = lambda: attention(q, k, v, q_scale, k_scale, v_scale, causal, sm_scale)
        if mode == "bwd":
            o = fn()
            do = torch.randn_like(o)
            fn = lambda: o.backward(do, retain_graph=True)
        ms = triton.testing.do_bench(fn, warmup=warmup, rep=rep)
    if provider == "flash":
        qkv = torch.randn((BATCH, N_CTX, 3, H, HEAD_DIM), dtype=dtype, device=device, requires_grad=True)
        fn = lambda: flash_attn_func(qkv, causal=causal)
        if mode == "bwd":
            o = fn()
            do = torch.randn_like(o)
            fn = lambda: o.backward(do, retain_graph=True)
        ms = triton.testing.do_bench(fn, warmup=warmup, rep=rep)
    flops_per_matmul = 2.0 * BATCH * H * N_CTX * N_CTX * HEAD_DIM
    total_flops = 2 * flops_per_matmul
    if causal:
        total_flops *= 0.5
    if mode == "bwd":
        total_flops *= 2.5  
    return total_flops / ms * 1e-9


if __name__ == "__main__":
    bench_flash_attention.run(print_data=True)
    test_op()
```

**Comment by Ezio-csm** [comment]
okey, thanks a lot.
